### ISSUE TITLE
Allow an organization to restrict signup to SSO / SAML on public SAAS

### ISSUE THREAD

**Issue by Dschoordsch** [ParabolInc/parabol#6095]
Relates to:
* #5512
* #4666
* #5663

From a support request:
As an organization (with enterprise tier and SAML) I want to restrict the email domains that allowed to be invited to
1) control license consumption
2) restrict access to the content

Restricting just the invite to certain domains would not solve their problem since it would still allow people to accept the invite with non-company addresses which they could still use for signup after leaving the company.

## Acceptance criteria
* [ ] add a field to Organization to enforceSAML
* [ ] if a new user accepts an invite to Org, they are required to signup with SAML if enforceSAML is set
* [ ] if an existing user tries to accept the invite and he's not a SAML user of that Org, they must create a new account

**Comment by mattkrick** [comment]
I expect this to be hard since we currently don't create an `AuthIdentity` when someone logs in with SAML.
Some proposed tips that may help the implementation:
- `enforceSAML` should be a foreign key that points to the PK of the `SAML` table. then, we can restrict on the SAML domains. Perhaps `enforcedSAMLId`?
- whenever someone logs in via SAML, we'll need to create an AuthIdentity on the user object for them if one doesn't exist. This should include the SAML primary key
- when someone tries to accept an invite, we'll have to check their User AuthIdentity & make sure the AuthIdentity.samlId === organization.enforcedSAMLId. This is important otherwise i could have created an unverified account before the domain became SAML-verified

**Comment by jordanh** [comment]
@mattkrick @Dschoordsch perhaps we divide this into multiple issues with @mattkrick's AC as the predecessor issue to @Dschoordsch's AC?

**Comment by drewhousman1010** [comment]
A [new customer](https://app.hubspot.com/contacts/3888472/contact/67283401) 🔒 writes: 

> We want to restrict our account to only SSO/OAuth. How do I do that? (I'm fine if it means we have to kick out the few people who used personal by mistake at first)

**Comment by mattkrick** [comment]
Since this is coming up soon I wanted to brainstorm some implementation.
I think Georg is right where the logic should exist on accepting an invitation.

AC
- [ ] `enableSAMLForDomain` includes an optional `lockedOrgIds` param that takes an array of orgIds. this field is set on the org.
- [ ] When I accept a teamMemberInvitation, the org gets loaded. the logic checks to see if `org.enforcedSAMLId` is set. if so, it pulls the SAML record. it compares SAML.domains to the domain of the email address accepting the invitation. if the domain isn't found, it throws an error saying this email cannot join this organization. it then gives a list of domains that can join. This is safe because they have been invited by an org member, so that means we're not leaking info to random users. 

the above is possible because once an org uses SAML they can't login with email/password

**Comment by mattkrick** [comment]
tangential, but for our own user story, we need a way to limit invites to a whitelist of approved tlds.
we had a case where we invited 300 people to our org by accident.
this is because it's too easy to invite randos to the org.
to fix this, we'll need an approved list of domains. it's not perfect, but it's a start

**Comment by mattkrick** [comment]
Let's see fi we can build something that fixes all these use cases:

### User Story 1

I own a company & I don't want people to accidentally log in using their personal gmail account. Instead, when they try to accept an invitation to the organization with their gmail account, they get an error saying they must log in using an email ending in xxx.

### User Story 2

I want to force all my employees to login via SAML. If they try to login with email/password, it tells them to login with SAML.

### User Story 3

I want to keep randos outside my org so I can control spending. If someone gets ahold of a mass invite token, it is only useful if their email ends in an approved domain. 

### User Story 4

I own a company & have restricted access to only the approved tlds. I bring in a contractor & he says he can't join my retro because of this. I want a 1-click button to allow this particular user to join via their gmail. When the contract has ended, I want to remove them from the org & that email from the whitelist.

### User Story 5

I don't trust my team. Only org admins should be able to send invites

Story 2 is already done. If someone tries to login using an email address tied to a SAML domain, the SAML popups is shown.
For the rest of the stories, each Org has a list of approved domains or email addresses. User Story 5 has nothing to do with the recipient, it's all about the receiver. I don't think we can solve for that use case without expanding the scope of the PR.

Using the `SAML.domains` as the source of that list is convenient but brittle. inevitably we're gonna get a customer who wants to make an exception & when that happens, we'll be SOL. It also doesn't scratch the itch for our own selfish use case (story 3).

I think the right way to do this is another DB table.

```ts
interface ApprovedOrganizationInvite {
    id: number
    orgId: string
    domain: string // domain or email
    addedByUserId: string // must be billing leader or superuser
    removedAt: Date // append-only, keep track of anyone who removes these
  }
```

to modify this, we'll need a graphql mutation on the public schema that requires org billing leader or superuser permission.
we could have 2 mutations: add/remove. or 1, e.g. modify(isRemove: false).
we also want this to be queryable: `organization { approvedDomains }`
we also need a rule that says if there are no rules in this table for an org, then we allow anyone.

AC
- [ ] TBD

**Comment by mattkrick** [comment]
acceptTeamInvitation can happen in parallel to signing up, logging in, or verifying email.

this is problematic because if i want to join parabol.co & that domain is restricted & i try to sign up with my gmail, then it'll create a new account with gmail, then fail joining the team. we'd like to avoid that because it creates an extra account on the server. the client will then have to decide whether to log them in or not, but regardless it's an extra, worthless account.

alternatively, if the invitation was already accepted, then this error should not prevent them from logging in.
in other words, signing up is conditionally dependent on the result of the invitation token. We may want to continue signing up. Or we may want to abort.

if this was anything other than sign up, i'd say "Good enough", but since signups are so darn critical to growth & virality, I think we need to get this right.

So, here's what I propose:
- LoginWithGoogle, LoginWithPassword, SignupWithPassword all accept an invitationToken. They pass it to a helper function that checks to make sure it's legit. if it's not legit, they exit early. 
- loginSAML is different because we don't want to pass an invitationToken to the IdP. However, the chances that someone logs in via SAML accidentally? pretty rare. So, for this we can handle the saml login first & then when it succeeds we can call acceptTeamInvitaiton. if that fails, we just don't log them in.
- verifyEmail stores the invitation token on the EmailVerification record. at the end of the mutation, it should validate the token. before joining the team. it looks like there's a bug in it where if the user already exists it ignores the invitationToken!
- acceptTeamInvitation mutation should look for that error & pop a snackbar. we can make it better later, but the scope is getting large here.  
- Is the user logged in?

**Comment by gcrickman** [comment]
A [user](https://app.hubspot.com/contacts/3888472/contact/76743701) asked about this today on our call regarding purchasing Parabol Enterprise.

**Comment by avivapinchas** [comment]
**Effort: L**
  [See the discussion](https://action.parabol.co/meet/dHOSUUpFO8/estimate/8) in Sprint Poker #​​1

  *Powered by [Parabol](https://www.parabol.co/resources/sprint-poker-meetings)*

**Comment by mattkrick** [comment]
this is finished & launching tomorrow
